File: scripts/sri.py
"""Prints SRI (sha384) hashes for every third-party script Nexus OS loads, and checks the
CDN sends CORS headers (required for integrity checks). Run: python scripts/sri.py
Use --check to compare against sri.js instead of printing (exits 1 on mismatch)."""
import base64, hashlib, re, sys, urllib.request

URLS = [
  "https://www.gstatic.com/firebasejs/10.14.1/firebase-app-compat.js",
  "https://www.gstatic.com/firebasejs/10.14.1/firebase-auth-compat.js",
  "https://www.gstatic.com/firebasejs/10.14.1/firebase-firestore-compat.js",
  "https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.min.js",
  "https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.worker.min.js",
  "https://cdnjs.cloudflare.com/ajax/libs/mammoth/1.6.0/mammoth.browser.min.js",
]

def fetch(u):
    req = urllib.request.Request(u, headers={"Origin": "https://example.github.io", "User-Agent": "nexus-sri"})
    with urllib.request.urlopen(req, timeout=60) as r:
        return r.read(), r.headers.get("Access-Control-Allow-Origin")

result = {}
for u in URLS:
    body, cors = fetch(u)
    result[u] = "sha384-" + base64.b64encode(hashlib.sha384(body).digest()).decode()
    print(("OK  " if cors else "NO-CORS ") + u.split("/")[-1].ljust(34), "%8d bytes" % len(body), "cors=" + str(cors), file=sys.stderr)

if "--check" in sys.argv:
    src = open("sri.js", encoding="utf-8").read()
    bad = [u for u, h in result.items() if h not in src]
    if bad:
        print("MISMATCH:", *bad, sep="\n  "); sys.exit(1)
    print("sri.js matches the live files"); sys.exit(0)
for u, h in result.items():
    print('  "%s": "%s",' % (u, h))

File: sri.js
/* ============================================================
   Nexus OS — Subresource Integrity for every third-party script.

   The browser refuses to run a script whose bytes don't match the hash, so a
   compromised or tampered CDN can't inject code into the portal (which handles
   resumes, tracker data and a short-lived Gmail token).

   All URLs are version-pinned, so the bytes never legitimately change.
   Re-generate after bumping a version:   python scripts/sri.py
   Verify against the live files:         python scripts/sri.py --check

   Google Identity Services (accounts.google.com/gsi/client) is deliberately not
   listed: Google serves it unversioned, so a pinned hash would break on their
   next release. It is restricted by the page's Content-Security-Policy instead.
   ============================================================ */
(function () {
  var HASH = {
    "https://www.gstatic.com/firebasejs/10.14.1/firebase-app-compat.js": "sha384-ZaR6mWzmJtrRibZ1Vm7SoHFr8OXjyAuGAXalGDKqbxFT18oi/z+oZLIRFkpeNor1",
    "https://www.gstatic.com/firebasejs/10.14.1/firebase-auth-compat.js": "sha384-I1LYojsZ5RM1cOda44Z2h42Qa6YfsQ1XkXxREnhp4ueYBR/4d1pG1K+NZM537Vsj",
    "https://www.gstatic.com/firebasejs/10.14.1/firebase-firestore-compat.js": "sha384-Ke0FJhH7LyRqDxZ0wt+/OXV38yfQVu7g9VPEEGjYmB4RVOY/ta04uecRhsMwT7V3",
    "https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.min.js": "sha384-/1qUCSGwTur9vjf/z9lmu/eCUYbpOTgSjmpbMQZ1/CtX2v/WcAIKqRv+U1DUCG6e",
    "https://cdnjs.cloudflare.com/ajax/libs/pdf.js/3.11.174/pdf.worker.min.js": "sha384-SnzOobpRMLXZ52iJvZm/C0fYw0OQemTXzTjIsdsfMcrCtCEe9qgzxTd3RSklO5x2",
    "https://cdnjs.cloudflare.com/ajax/libs/mammoth/1.6.0/mammoth.browser.min.js": "sha384-nFoSjZIoH3CCp8W639jJyQkuPHinJ2NHe7on1xvlUA7SuGfJAfvMldrsoAVm6ECz"
  };

  var cache = {};

  /* Load a script, verifying its hash when we have one. Repeated calls share one load. */
  function script(url) {
    if (cache[url]) return cache[url];
    cache[url] = new Promise(function (res, rej) {
      var s = document.createElement("script");
      s.src = url; s.async = true;
      if (HASH[url]) { s.integrity = HASH[url]; s.crossOrigin = "anonymous"; }
      s.onload = function () { res(); };
      s.onerror = function () {
        delete cache[url];
        rej(new Error("Couldn't load " + url.split("/").pop() + " - check your connection" + (HASH[url] ? " (or the file failed its integrity check)" : "") + "."));
      };
      document.head.appendChild(s);
    });
    return cache[url];
  }

  /* fetch() a text file with the same integrity guarantee (used for the PDF worker) */
  function text(url) {
    var init = HASH[url] ? { integrity: HASH[url], mode: "cors" } : {};
    return fetch(url, init).then(function (r) {
      if (!r.ok) throw new Error("HTTP " + r.status);
      return r.text();
    });
  }

  window.NXSRI = { HASH: HASH, script: script, text: text };
})();
